### PR TITLE
new CommonSourceDatasetMeasurements and MeasurementPairs classes crea…

### PR DESCRIPTION
**Pull request by GorkaLeonie**
new CommonSourceDatasetMeasurements and MeasurementPairs classes created to overcome overflow of if-else statements

### CODE REVIEW

File: lrbenchmark/data/dataset.py [hunk 1]
```diff
@@ -83,30 +83,26 @@ def pop(self, fraction: float, seed: int = None) -> 'Dataset':
         raise NotImplementedError
 
 
-class CommonSourceKFoldDataset(Dataset, ABC):
-    def __init__(self, n_splits: Optional[int], measurements: Optional[List[Measurement]] = None,
-                 measurement_pairs: Optional[List[MeasurementPair]] = None):
+class CommonSourceDataset(Dataset, ABC):
```
Comment (rolfypma-nfi) [inline lrbenchmark/data/dataset.py:86]: for general discussion; i dont think common source is a property of a dataset, but of how you handle the dataset. suggest to remove this in-between class (other PR)

File: lrbenchmark/data/dataset.py [hunk 2]
```diff
@@ -83,30 +83,26 @@ def pop(self, fraction: float, seed: int = None) -> 'Dataset':
...
+    def __init__(self, n_splits: Optional[int]):
         super().__init__()
         self.n_splits = n_splits
```
Comment (rolfypma-nfi) [inline lrbenchmark/data/dataset.py:89]: why is n_splits a property of the object, and not a parameter of calling get_splits?

File: lrbenchmark/data/dataset.py [hunk 3]
```diff
@@ -172,50 +245,20 @@ def get_splits_measurement_pairs(self,
...
-            yield [CommonSourceKFoldDataset(n_splits=None,
-                                            measurements=list(map(lambda i: self.measurements[i], split_idx))) for
-                   split_idx in split]
-
-    def get_x_y_pairs(self,
-                      seed: Optional[int] = None,
-                      pairing_function: Optional[Callable] = partial(InstancePairing,
-                                                                     different_source_limit='balanced'),
-                      transformer: Optional[Callable] = AbsDiffTransformer) -> XYType:
+    def get_x_y_pairs(self, transformer: Optional[Callable] = AbsDiffTransformer) -> XYType:
```
Comment (rolfypma-nfi) [inline lrbenchmark/data/dataset.py:261]: it's very confusing to have 2 get_x_y functions. Can we make it one function that checks whether any computation needs to be done? I would suggest leaving the transformation to whoever call this, and let the dataset just provide the data

File: lrbenchmark/data/dataset.py [hunk 4]
```diff
@@ -225,26 +268,20 @@ def get_x_y_pairs(self,
...
+            # the shape of the measurement values should be (m, f,2), with m=number of pairs, f=number of features
+            # and 2 values (for the two measurements), to be compatible with the transformation function
+            X_pairs = np.array([mp.get_measurement_values() for mp in self.measurement_pairs])
+            y_pairs = np.array([mp.is_same_source for mp in self.measurement_pairs])
             X_pairs = transformer().transform(X_pairs)
             return X_pairs, y_pairs
 
 
-class XTCDataset(CommonSourceKFoldDataset):
+class XTCDataset(CommonSourceDataset):
```
Comment (rolfypma-nfi) [inline lrbenchmark/data/dataset.py:284]:
```suggestion
class XTCDataset(CommonSourceDatasetMeasurements):
```

**Review by rolfypma-nfi — CHANGES_REQUESTED** [review]
I think splitting these two datasets makes sense!

File: lrbenchmark/data/dataset.py (hunk 2 above)
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:89]: fixed by removing n_splits all together and hard-coding this to the value 1 for the splits.

File: lrbenchmark/data/dataset.py [hunk 5]
```diff
@@ -83,30 +83,26 @@ def pop(self, fraction: float, seed: int = None) -> 'Dataset':
...
+            self.load()
+
     @property
     def source_ids(self) -> Set[int]:
-        if self.measurements:
-            return set([m.source.id for m in self.measurements])
-        else:
-            return set(chain.from_iterable(
-                [[mp.measurement_a.source.id, mp.measurement_b.source.id] for mp in self.measurement_pairs]))
+        return set([m.source.id for m in self.measurements])
```
[the remark this answers is not in the document]
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:105]: changed

File: lrbenchmark/data/dataset.py [hunk 6]
```diff
@@ -83,30 +83,26 @@ def pop(self, fraction: float, seed: int = None) -> 'Dataset':
...
 
     def get_x_measurement(self) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:107]: adopted

File: lrbenchmark/data/dataset.py [hunk 7]
```diff
@@ -117,20 +113,100 @@ def get_y_measurement(self) -> np.ndarray:
     def get_x_y_measurement(self) -> XYType:
```
[the remark this answers is not in the document]
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:113]: adopted

File: lrbenchmark/data/dataset.py [hunk 8]
```diff
@@ -117,20 +113,100 @@ def get_y_measurement(self) -> np.ndarray:
...
+        """
+        This function splits the measurements in a dataset into two splits, as specified by the
+        provided parameters.
+
+        :param stratified: boolean that indicates whether the dataset should be split while preserving the ratio of
+                           classes in y in both splits.
+        :param group_by_source: boolean that indicates whether the dataset should be split along group lines, ensuring
+                                each group to be in only a single split.
+        :param train_size: size of the train set. Can be a float, to indicate a fraction, or an integer to indicate an
+                           absolute amount of measurements, measurement pairs or groups in each split.  If not
```
[the remark this answers is not in the document]
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:128]: adopted

File: lrbenchmark/data/dataset.py [hunk 9]
```diff
@@ -117,20 +113,100 @@ def get_y_measurement(self) -> np.ndarray:
...
+                           specified, is the complement of the test_size.
+        :param test_size: size of the test set. Can be a float, to indicate a fraction, or an integer to indicate an
+                          absolute amount of measurements, measurement pairs or groups in each split. If not specified,
```
[the remark this answers is not in the document]
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:131]: adopted

File: lrbenchmark/data/dataset.py [hunk 10]
```diff
@@ -117,20 +113,100 @@ def get_y_measurement(self) -> np.ndarray:
...
+                          is the complement of the train_size.
+        :param seed: seed to ensure repeatability of the split
+        """
+        yield from self.get_splits_measurements(group_by_source, stratified, train_size, test_size, seed)
```
[the remark this answers is not in the document]
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:135]: fixed

File: lrbenchmark/data/dataset.py [hunk 11]
```diff
@@ -149,16 +139,15 @@ def get_splits_measurements(self,
             raise ValueError('It is not possible to split the dataset stratified, when using measurements')
 
         if not group_by_source:
-            s = ShuffleSplit(n_splits=self.n_splits, random_state=seed, train_size=train_size, test_size=test_size)
+            s = ShuffleSplit(n_splits=1, random_state=seed, train_size=train_size, test_size=test_size)
```
Comment (rolfypma-nfi) [inline lrbenchmark/data/dataset.py:142]: n_splits=1 may not be the expected behaviour :)
make  n_splits a parameter of the function?

File: lrbenchmark/data/dataset.py [hunk 12]
```diff
@@ -149,16 +139,15 @@ def get_splits_measurements(self,
...
         else:
-            s = GroupShuffleSplit(n_splits=self.n_splits, random_state=seed, train_size=train_size, test_size=test_size)
+            s = GroupShuffleSplit(n_splits=1, random_state=seed, train_size=train_size, test_size=test_size)
             source_ids = [m.source.id for m in self.measurements]
 
-        for split in s.split(self.measurements, groups=source_ids):
-            yield [CommonSourceDatasetMeasurements(n_splits=None,
-                                                   measurements=list(map(lambda i: self.measurements[i], split_idx)))
-                   for split_idx in split]
+        split = s.split(self.measurements, groups=source_ids)[0]
```
Comment (rolfypma-nfi) [inline lrbenchmark/data/dataset.py:148]: why only take the first split?

File: lrbenchmark/data/dataset.py [hunk 13]
```diff
@@ -191,17 +180,14 @@ def __init__(self, n_splits: Optional[int], measurement_pairs: Optional[List[Mea
...
         return np.array([mp.get_x() for mp in self.measurement_pairs])
 
-    def get_y_mp(self) -> np.ndarray:
+    def get_y(self) -> np.ndarray:
         return np.array([mp.get_y() for mp in self.measurement_pairs])
 
-    def get_x_y_mp(self) -> XYType:
-        return self.get_x_mp(), self.get_y_mp()
-
     def get_splits(self, stratified: bool = False, group_by_source: bool = False,
```
Comment (rolfypma-nfi) [inline lrbenchmark/data/dataset.py:191]: dont forget to remove this one

File: lrbenchmark/data/dataset.py [hunk 14]
```diff
@@ -83,30 +83,26 @@ def pop(self, fraction: float, seed: int = None) -> 'Dataset':
...
         raise NotImplementedError
 
+
+class CommonSourceDatasetMeasurements(CommonSourceDataset):
+    def __init__(self, n_splits: Optional[int], measurements: Optional[List[Measurement]] = None):
+        super().__init__(n_splits)
+        self.measurements = measurements
+
+        if self.measurements is None:
+            self.load()
```
[the remark this answers is not in the document]
Comment (rolfypma-nfi) [reply lrbenchmark/data/dataset.py:100]: in other PR

[the remark this answers is not in the document]
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:100]: in other PR

File: lrbenchmark/data/dataset.py [hunk 15]
```diff
@@ -117,20 +113,100 @@ def get_y_measurement(self) -> np.ndarray:
...
 
-    def get_x_measurement_pair(self) -> np.ndarray:
+    def get_splits(self, stratified: bool = False, group_by_source: bool = False,
+                   train_size: Optional[Union[float, int]] = 0.8, test_size: Optional[Union[float, int]] = 0.2,
+                   seed: int = None) -> Iterable[Dataset]:
+        """
+        This function splits the measurements in a dataset into two splits, as specified by the
+        provided parameters.
+
+        :param stratified: boolean that indicates whether the dataset should be split while preserving the ratio of
```
[the remark this answers is not in the document]
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:122]: in other PR

File: lrbenchmark/data/dataset.py [hunk 16]
```diff
@@ -117,20 +113,100 @@ def get_y_measurement(self) -> np.ndarray:
     def get_x_y_measurement(self) -> XYType:
         return self.get_x_measurement(), self.get_y_measurement()
 
-    def get_x_measurement_pair(self) -> np.ndarray:
+    def get_splits(self, stratified: bool = False, group_by_source: bool = False,
+                   train_size: Optional[Union[float, int]] = 0.8, test_size: Optional[Union[float, int]] = 0.2,
```
[the remark this answers is not in the document]
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py]: in other PR

File: lrbenchmark/data/dataset.py [hunk 17]
```diff
@@ -117,20 +113,100 @@ def get_y_measurement(self) -> np.ndarray:
...
+        return X_pairs, y_pairs
+
+
+class CommonSourceDatasetMeasurementPairs(CommonSourceDataset):
+    def __init__(self, n_splits: Optional[int], measurement_pairs: Optional[List[MeasurementPair]] = None):
+        super().__init__(n_splits)
+        self.measurement_pairs = measurement_pairs
+
+        if self.measurement_pairs is None:
+            self.load()
```
[the remark this answers is not in the document]
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:178]: in other PR

File: lrbenchmark/data/dataset.py (hunk 11 above)
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:142]: adopted

File: lrbenchmark/data/dataset.py (hunk 12 above)
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:148]: resolved

File: lrbenchmark/data/dataset.py [hunk 18]
```diff
@@ -117,20 +113,100 @@ def get_y_measurement(self) -> np.ndarray:
...
+        else:
+            s = GroupShuffleSplit(n_splits=self.n_splits, random_state=seed, train_size=train_size, test_size=test_size)
+            source_ids = [m.source.id for m in self.measurements]
+
+        for split in s.split(self.measurements, groups=source_ids):
+            yield [CommonSourceDatasetMeasurements(n_splits=None,
+                                                   measurements=list(map(lambda i: self.measurements[i], split_idx)))
+                   for split_idx in split]
+
+    def get_x_y_pairs(self,
```
[the remark this answers is not in the document]
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py:164]: task created in backlog

File: lrbenchmark/data/dataset.py (hunk 3 above)
Comment (GorkaLeonie) [reply lrbenchmark/data/dataset.py]: task created in backlog

File: lrbenchmark/data/dataset.py [hunk 19]
```diff
@@ -83,138 +91,169 @@ def pop(self, fraction: float, seed: int = None) -> 'Dataset':
...
-        else:
-            s = GroupShuffleSplit(n_splits=self.n_splits, random_state=seed, train_size=train_size, test_size=test_size)
-            source_ids = [m.source.id for m in self.measurements]
-
-        for split in s.split(self.measurements, groups=source_ids):
-            yield [CommonSourceKFoldDataset(n_splits=None,
-                                            measurements=list(map(lambda i: self.measurements[i], split_idx))) for
-                   split_idx in split]
-
     def get_x_y_pairs(self,
```
Comment (rolfypma-nfi) [inline lrbenchmark/data/dataset.py:256]: deze dan ook gewoon get_x_y?

**Review by rolfypma-nfi — APPROVED** [review]

**Pull request merged**